pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -687,22 +687,24 @@ int ConsumerImpl::getNumOfPrefetchedMessages() const {
     return incomingMessages_.size();
 }
 
-Result ConsumerImpl::getConsumerStats(BrokerConsumerStats& brokerConsumerStats, int partitionIndex) {
+void ConsumerImpl::getConsumerStatsAsync(BrokerConsumerStatsCallback callback, int partitionIndex) {
     if (partitionIndex != -1) {
         LOG_WARN(getName() << "Ignoring the partitionIndex since the topic is not partitioned")
     }
 
-    if (!isOpen()) {
+    Lock lock(mutex_);
+    if (state_ != Ready) {
         LOG_ERROR(getName() << "Client connection is not open, please try again later.")
-        return ResultConsumerNotInitialized;
+        lock.unlock();
+        return callback(ResultConsumerNotInitialized, BrokerConsumerStats());
     }
 
     if (brokerConsumerStats_.isValid()) {
         LOG_DEBUG(getName() << "Serving data from cache");
-        brokerConsumerStats = brokerConsumerStats_;
-        return ResultOk;
+        lock.unlock();
+        return callback(ResultOk, brokerConsumerStats_);
```

Comment:
this callback isn't safe as you just dropped the lock.  need to copy ```brokerConsumerStats_``` before unlocking.